src/lib.rs

```diff
@@ -1,59 +1,74 @@
-use serde_json::{Value};
+use serde_json::{Map, Value};
+use thiserror::Error;
 
-pub fn hydrate(base: Value, item: &mut Value) {
-    if item.is_null(){
-        *item = base.clone();
-    }
-    else if item.is_object() {
-        if item.as_object() == base.as_object(){
-            return;
+const MAGIC_MARKER: &str = "𒍟※";
+
+pub trait Hydrate {
+    fn hydrate(&mut self, base: Self) -> Result<(), Error>;
+}
+
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error("type mismatch")]
+    TypeMismatch(Value, Value),
+}
+
+impl Hydrate for Value {
+    fn hydrate(&mut self, base: Self) -> Result<(), Error> {
+        match self {
+            Value::Object(item) => match base {
+                Value::Object(base) => item.hydrate(base),
+                _ => Err(Error::TypeMismatch(self.clone(), base)),
+            },
+            Value::Array(item) => match base {
+                Value::Array(base) => item.hydrate(base),
+                _ => Err(Error::TypeMismatch(self.clone(), base)),
+            },
+            _ => Ok(()),
         }
-        if let Value::Object(item) = item {
-            if let Value::Object(ref base) = base {
-                for (key, baseval) in base {
-                    if item.get(key).is_some(){
-                        let itemval = item.get_mut(key).unwrap();
-                        if itemval.as_str() == Some("𒍟※" ){
-                            item.remove(key);
-                        }
-                        else {
-                            hydrate(baseval.clone(), itemval);
-                        }
-
-                    }
-                    else {
-                        item.entry(key).or_insert(baseval.clone());
-                    }
-
-                }
-            }
+    }
+}
+
+impl Hydrate for Vec<Value> {
+    fn hydrate(&mut self, base: Self) -> Result<(), Error> {
+        for (item, base) in self.iter_mut().zip(base.into_iter()) {
+            item.hydrate(base)?;
         }
+        Ok(())
     }
-    else if item.is_array(){
-        if let Value::Array(item) = item {
-            if let Value::Array(ref base) = base {
-                if item.len() == base.len(){
-                    println!("item len == base len");
-                    for i in 0..item.len(){
-                        hydrate(base[i].clone(), &mut item[i]);
-                    }
-                }
```

Comment:
It felt a little strange to just silently skip arrays that aren't the same length. To me, it should either be an error (which is how I originally implemented it, until I saw that there was a test that expected it to succeed w/ unequal length arrays) or it should still do hydration but stop when it runs out of values from either `item` or `base` (how it's currently implemented.

Reply:
I think we may want to add a strict flag to the function that if on would error in a situation like this if off, just defaults to returning the item value.